runner/src/server/plugins/engine/components/ClientSideFileUploadField.ts

```diff
@@ -0,0 +1,83 @@
+import { FormComponent } from "./FormComponent";
+import { FormData, FormSubmissionErrors } from "../types";
+import { ClientSideFileUploadFieldComponent } from "@xgovformbuilder/model";
+import { FormModel } from "server/plugins/engine/models";
+import { ClientSideFileUploadFieldViewModel } from "server/plugins/engine/components/types";
+import joi from "joi";
+export class ClientSideFileUploadField extends FormComponent {
+  options: ClientSideFileUploadFieldComponent["options"];
+  schema: ClientSideFileUploadFieldComponent["schema"];
+
+  constructor(def: ClientSideFileUploadFieldComponent, model: FormModel) {
+    super(def, model);
+    this.options = def.options;
+    this.schema = def.schema;
+  }
+
+  getFormSchemaKeys() {
+    return {
+      [this.name]: joi.allow("").allow(null),
+      [`${this.name}__filename`]: joi.string().optional(),
+      [`${this.name}__delete[]`]: joi.string().optional(),
+    };
+  }
+
+  getAdditionalValidationFunctions(): Function[] {
+    return [
+      async (request, viewModel) => {
+        const { uploadService, cacheService } = request.services([]);
+        const state = await cacheService.getState(request);
+        const applicationId = state.metadata?.application_id ?? "";
+        const { id, path } = request.params as any;
+        const componentKey = viewModel.components.find(
+          (c) => c.type === "ClientSideFileUploadField"
+        ).model.id;
+        const key = `${applicationId}/${id}/${path}/${componentKey}`;
+
+        // we wait an arbitrary amount of 1 second here, because of race conditions.
+        await new Promise((resolve) => setTimeout(resolve, 1000));
```

Comment:
I don't like this, but I found if I quickly deleted a file and submitted at the same time, there was a race condition where the validation would pass and then the file would be deleted afterwards - this small time period is to account for that scenario.  Open ears on a more suitable option for this....